### PR TITLE
add commands to support China/Tencent user

### PR DESCRIPTION
**Pull request by hkbarton**
## What has been implemented?

Add commands to support Tencent/Chinese users. It includes:

- add reference of `@serverless/tencent-platform-client`
- update IS_IN_CHINA flag, so if the user manually set `process.env.SERVERLESS_PLATFORM_VENDOR === 'aws'` we won't treat them as China user, or if the user manually set `process.env.SERVERLESS_PLATFORM_VENDOR === 'tencent'` we'll treat them as China user regardless the zimezone
- based on IS_IN_CHINA flag, dynamically require different command set
- add command sets for China/Tencent user

### CODE REVIEW

**Comment by eahefnawy** [comment]
@hkbarton could you fix the conflicts and reopen this against `master`? I just merged `cloud` into `master`

**Comment by medikoo** [comment]
@hkbarton I've already changed the `base` to `master` on a PR

File: src/cli/commands-cn/registry.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  cli.log()
+
+  cli.log(`Featured Components:`)
+  cli.log()
+  cli.log(`  express - https://github.com/serverless-components/express`)
+  cli.log(`  website - https://github.com/serverless-components/website`)
+  cli.log(`  aws-lambda - https://github.com/serverless-components/aws-lambda`)
+  cli.log(`  aws-dynamodb - https://github.com/serverless-components/aws-dynamodb`)
+  cli.log(`  aws-iam-role - https://github.com/serverless-components/aws-iam-role`)
+  cli.log(`  aws-lambda-layer - https://github.com/serverless-components/aws-lambda-layer`)
```
Comment (medikoo) [inline src/cli/commands-cn/registry.js:121]: I believe there should be no AWS references

File: src/cli/commands-cn/utils.js [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+  refreshToken,
+  listTenants
+} = require('@serverless/platform-sdk')
+const { utils: platformUtils } = require('@serverless/tencent-platform-client')
+
+/**
+ * Wait for a number of miliseconds
+ * @param {*} wait
+ */
+const sleep = async (wait) => new Promise((resolve) => setTimeout(() => resolve(), wait))
```
Comment (medikoo) [inline src/cli/commands-cn/utils.js:28]: It'll be great if utils are not duplicate.

Generic utils (as `sleep`) are kept in outer `../utils.js` and here we should have just Tencent specific ones.

It'll make review process easier

File: src/cli/utils.js [hunk 3]
```diff
@@ -350,7 +350,11 @@ const legacyLoadComponentConfig = (directoryPath) => {
 // Same as internal Tencent check:
 // https://github.com/serverless-tencent/serverless-tencent-tools/blob/3c1cabbdb21c0b3ba37248b9c2f609ec552bf8fc/sdk/others/isInChina.js#L12
 const IS_IN_CHINA =
-  new Date().getTimezoneOffset() == -480 || String(process.env.LC_CTYPE).includes('zh_CN')
+  (Boolean(
+    new Date().getTimezoneOffset() == -480 || String(process.env.LC_CTYPE).indexOf('zh_CN')
+  ) &&
```
Comment (medikoo) [inline src/cli/utils.js:355]: This check should be same as in master (above change actually reverts important bug fix https://github.com/serverless/components/pull/568 )

File: src/cli/commands-cn/info.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  // format last action for better UX
+  const lastActionAgo = moment(instance.lastActionAt).fromNow()
+
+  const dashboardUrl = utils.getInstanceDashboardUrl(instanceYaml)
+
+  // show the most important information, and link to the dashboard
+  cli.log(`${chalk.grey('Status:')}       ${instance.instanceStatus}`)
+  cli.log(`${chalk.grey('Last Action:')}  ${instance.lastAction} (${lastActionAgo})`)
+  cli.log(`${chalk.grey('Deployments:')}  ${instance.instanceMetrics.deployments}`)
+  cli.log(`${chalk.grey('More Info:')}    ${dashboardUrl}`)
```
[the remark this answers is not in the document]
Comment (hkbarton) [reply src/cli/commands-cn/info.js:67]: done

File: src/cli/commands-cn/registry.js (hunk 1 above)
Comment (hkbarton) [reply src/cli/commands-cn/registry.js:121]: yes, updated

File: src/cli/commands-cn/utils.js (hunk 2 above)
Comment (hkbarton) [reply src/cli/commands-cn/utils.js:28]: yes, I haven't fully cleaned up utils, let me do that

File: src/cli/utils.js (hunk 3 above)
Comment (hkbarton) [reply src/cli/utils.js:355]: done

File: src/cli/commands-cn/utils.js [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  const envVars = initEnvs(stage)
+
+  // Known Provider Environment Variables and their SDK configuration properties
+  const providers = {}
+
+  // AWS
+  providers.aws = {}
+  providers.aws.AWS_ACCESS_KEY_ID = 'accessKeyId'
+  providers.aws.AWS_SECRET_ACCESS_KEY = 'secretAccessKey'
+  providers.aws.AWS_REGION = 'region'
```
Comment (medikoo) [inline src/cli/commands-cn/utils.js:184]: I think here it's safe to stick just to Tencent (?)

Comment (hkbarton) [reply src/cli/commands-cn/utils.js:184]: ya, it's not needed, let me remove those.

**Comment by eahefnawy** [comment]
Thanks for the PR @hkbarton 🙌 

Didn't have time to review, but could we put a hold on merging this until after launch today?

**Comment by hkbarton** [comment]
> Thanks for the PR @hkbarton 🙌
> 
> Didn't have time to review, but could we put a hold on merging this until after launch today?

sure, let's lunch aws first and then merge this.

**Comment by ac360** [comment]
This can be merged now.  Good luck!

**Comment by eahefnawy** [comment]
@hkbarton with this PR, do we have Tencent support for all components commands? Or is there something still missing?

**Comment by hkbarton** [comment]
In this PR:

### What's support for China User

1. `info` command to show instance-related information
2. registry-related commands
  2.1 `sls registry [component name]` to list detail and available versions of the given component
  2.2 `sls registry` to list featured component
  2.3 `sls registry publish` to publish a component to registry
3. deployment-related commands
  3.1 `sls` or `sls deploy` to deploy an instance
  3.2 `sls dev` to enter the dev mode. Dev mode in China supports: log forwarding; using chrome inspector to remote debug NodeJs10.x and above app
4. User onboarding flow:
  4.1 `sls create` to create a project from a template
  4.2 Once merge @medikoo 's chance we'll have templates and onboarding experience for the user in China.

### What's not supported

1. `--debug` option to show component logs, this is not supported for now and will be supported after 4/8 lunch
2. `sls login/logout` is not needed for the user in China as we discussed before


@eahefnawy

File: src/cli/utils.js [hunk 6]
```diff
@@ -347,7 +348,31 @@ const legacyLoadComponentConfig = (directoryPath) => {
...
+  const stageEnvFilePath = path.join(process.cwd(), `.env.${stage}`)
+
+  // Load environment variables via .env file
+  if (stage && fileExistsSync(stageEnvFilePath)) {
+    envVars = dotenv.config({ path: path.resolve(stageEnvFilePath) }).parsed || {}
+  } else if (fileExistsSync(defaultEnvFilePath)) {
+    envVars = dotenv.config({ path: path.resolve(defaultEnvFilePath) }).parsed || {}
+  }
+  return envVars
+}
```
Comment (medikoo) [inline src/cli/utils.js:363]: Why this is introduced in a util?

I believe that this should be ensured in CLI intialization, and in utils we should just assume that env variables are resolved

File: src/cli/utils.js [hunk 7]
```diff
@@ -405,15 +405,6 @@ const legacyLoadComponentConfig = (directoryPath) => {
...
 
-const IS_IN_CHINA = (() => {
-  if (process.env.SLS_GEO_LOCATION === 'cn') {
-    return true
-  }
-  return new Intl.DateTimeFormat('en', { timeZoneName: 'long' })
-    .format()
-    .includes('China Standard Time')
-})()
-
```
Comment (medikoo) [inline src/cli/utils.js:416]: I believe this util should not be removed (I rely on it in https://github.com/serverless/components/pull/567)

Comment (hkbarton) [reply src/cli/utils.js:416]: can we use
```
const { utils: platformUtils } = require('@serverless/tencent-platform-client')
if (platformUtils.isChinaUser()) {
   ....
}
```
instead?

**Pull request merged**